domain/relation/state/relation.go

```diff
@@ -481,6 +483,389 @@ WHERE  relation_id = $getRelation.relation_id
 	}, nil
 }
 
+// EnterScope indicates that the provided unit has joined the relation.
+//
+// The following error types can be expected to be returned:
+//   - [relationerrors.RelationNotFound] If the relation cannot be found.
+//   - [relationerrors.UnitNotFound] If no unit by the given name can be found
+//   - [relationerrors.PotentialRelationUnitNotValid] If the unit entering scope
+//     is a subordinate and its endpoint has scope charm.ScopeContainer, but the
+//     principal application of the unit is not the application in the relation.
+func (st *State) EnterScope(
+	ctx context.Context,
+	relationUUID corerelation.UUID,
+	unitName unit.Name,
+) error {
+	db, err := st.DB()
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	type getUnit struct {
+		UnitUUID unit.UUID `db:"uuid"`
+		UnitName unit.Name `db:"name"`
+	}
+	unitArgs := getUnit{
+		UnitName: unitName,
+	}
+	getUnitStmt, err := st.Prepare(`
+SELECT &getUnit.*
+FROM   unit
+WHERE  name = $getUnit.name
+`, unitArgs)
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		// Get the UUID of the unit entering scope.
+		err = tx.Query(ctx, getUnitStmt, unitArgs).Get(&unitArgs)
+		if errors.Is(err, sqlair.ErrNoRows) {
+			return relationerrors.UnitNotFound
+		} else if err != nil {
+			return errors.Capture(err)
+		}
+
+		// Check the unit can enter scope in this relation.
+		err := st.checkUnitCanEnterScope(ctx, tx, relationUUID, unitArgs.UnitUUID)
+		if err != nil {
+			return errors.Errorf("checking unit valid in relation: %w", err)
+		}
+
+		// Upsert the row recording that the unit has entered scope.
+		err = st.upsertRelationUnit(ctx, tx, relationUUID, unitArgs.UnitUUID)
+		if err != nil {
+			return errors.Capture(err)
+		}
+		return err
+	})
+	return errors.Capture(err)
+}
+
+// checkUnitCanEnterScope checks that the unit can enter scope in the given
+// relation.
```

Comment:
todo: check that the unit and relation are alive, units/relations which dying and dead cannot enter scope.